ext/mbstring/mbstring.c

```diff
@@ -2282,6 +2289,280 @@ PHP_FUNCTION(mb_output_handler)
+	if(params->split_length == mb_chunk_size) { /* if current chunk size reached defined chunk size */
+		mbfl_convert_filter_flush(params->next_filter);/* concatenate separate decoded chars to the solid string */
+		mbfl_memory_device *device = (mbfl_memory_device *)params->next_filter->data; /* chars container */
+		mbfl_string *chunk = params->result_string;
+		mbfl_memory_device_result(device, chunk); /* make chunk */
+		add_next_index_stringl(params->return_value, chunk->val, chunk->len); /* add chunk to the array */
+		efree(chunk->val);
+		params->mb_chunk_size = 0; /* reset mb_chunk size */
+	}
+	return 0;
+}
+
+PHP_FUNCTION(mb_str_split)
+{
+	zend_string *str, *encoding = NULL;
+	size_t mb_len, chunks, chunk_len;
+	char unsigned const *last, *p;
+	mbfl_string string, result_string;
+	zend_long split_length = 1;
+
+	ZEND_PARSE_PARAMETERS_START(1, 3)
+		Z_PARAM_STR(str)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_LONG(split_length)
+		Z_PARAM_STR(encoding)
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (split_length <= 0) {
+		php_error_docref(NULL, E_WARNING, "The length of each segment must be greater than zero");
+		RETURN_FALSE;
+	}
+
+	/* fill mbfl_string structure */
+	string.no_language = MBSTRG(language);
+	string.encoding = php_mb_get_encoding(encoding);
+	if (!string.encoding) {
+		RETURN_FALSE;
+	}
+
+	p = (char unsigned *)ZSTR_VAL(str); /* string cursor pointer */
+	string.val = (char unsigned *)p;
+	string.len = ZSTR_LEN(str);
+	last = string.val + string.len;
+
+
+	/*
+	 * +----------------------------------------------------------------------+
+	 * |                           4 scenarios                                |
+	 * +----------------------------------------------------------------------+
+	 */
+	const mbfl_encoding *mbfl_encoding = string.encoding;
+
+	/*
+	 * +----------------------------------------------------------------------+
+	 * | first scenario: "fixed width encodings"                              |
+	 * +----------------------------------------------------------------------+
+	 */
+	if (mbfl_encoding->flag & MBFL_ENCTYPE_SBCS) { /* 1 byte */
+		mb_len = string.len;
+		chunk_len = (size_t)split_length; /* chunk length in bytes */
+
+	} else if (mbfl_encoding->flag & (MBFL_ENCTYPE_WCS2BE | MBFL_ENCTYPE_WCS2LE )) { /* 2 bytes */
+		mb_len = string.len>>1; /* eq. string.len / 2 String length in chars */
+		chunk_len = split_length<<1; /* eq. split_length * 2 chunk length in bytes */
+
+	} else if (mbfl_encoding->flag & (MBFL_ENCTYPE_WCS4BE | MBFL_ENCTYPE_WCS4LE )) { /* 4 bytes */
+		mb_len = string.len>>2; /* eq. string.len / 4 */
+		chunk_len = split_length<<2; /* eq. split_length * 4 */
+
+	/*
+	 * +----------------------------------------------------------------------+
+	 * | second scenario: "2- or 4-bytes width encodings UTF-16LE UTF-16BE    |
+	 * +----------------------------------------------------------------------+
+	 */
+	} else if (mbfl_encoding->flag & ( MBFL_ENCTYPE_MWC2LE | MBFL_ENCTYPE_MWC2BE )) {
```

Comment:
If we want to have this kind of optimization, it should be part of libmbfl, rather than only this one function.

Personally, I don't think this is necessary. Thankfully UTF-16 is not an important encoding in PHP and it's not an issue if it does not have a very fast implementation.

Reply:
I could try to implement it for the libmbfl if more experienced developers approve of this approach. I mean 2 relatively large tables of 65k each.

Reply:
This way is at least twice faster than libmfl way.

Reply:
Do i need to create new RFC for this?